tests/test_fipa_end2end.py

```diff
@@ -0,0 +1,307 @@
+    thread: Thread
+
+    @classmethod
+    def start(cls, multi_addr: Address):
+        """Configure agent and start it."""
+        cls.setup_class()
+
+        result = cls.invoke(
+            "config",
+            "set",
+            "--type",
+            "dict",
+            "vendor.fetchai.connections.p2p_libp2p.config",
+            f"""{{
+          "delegate_uri": "127.0.0.1:11001",
+          "entry_peers": ["{multi_addr}"],
+          "local_uri": "127.0.0.1:9001",
+          "log_file": "libp2p_node.log",
+          "public_uri": "127.0.0.1:9001"
+        }}""",
+        )
+        assert result.exit_code == 0
+
+        result = cls.invoke("get-address", "fetchai",)
+        assert result.exit_code == 0
+        addr = result.stdout.strip()
+        cls._run()
+        return addr
+
+    @classmethod
+    def _run(cls):
+        """Run agent."""
+        builder = AEABuilder.from_aea_project(cls._get_cwd())
+        skill_context = SkillContext()
+        handler = SellerHandler(name="handler", skill_context=skill_context)
+        test_skill = Skill(
+            SkillConfig(name="test_skill", author="fetchai"),
+            skill_context=skill_context,
+            handlers={"handler": handler},
+            behaviours={},
+        )
+        builder.add_component_instance(test_skill)
+        with cd(cls._get_cwd()):
+            cls.agent = builder.build()
+            skill_context.set_agent_context(cls.agent.context)
+
+        cls.thread = Thread(target=cls.agent.start, daemon=True)
+        cls.thread.start()
+        wait_for_condition(lambda: cls.agent.is_running, timeout=20)
+
+    @classmethod
+    def stop(cls):
+        """Stop agent and tear down."""
+        cls.agent.stop()
+        cls.thread.join(20)
+        cls.teardown_class()
+
+
+class TestFipaEnd2End(Base):
+    """Test that echo skill works."""
+
+    def test_run(self):
+        """Run the echo skill sequence."""
+        result = self.invoke(
+            "get-multiaddress",
+            "fetchai",
+            "-c",
+            "-i",
+            "fetchai/p2p_libp2p:0.21.0",
+            "-u",
+            "public_uri",
+        )
+        assert result.exit_code == 0
+        multi_addr = result.stdout.strip()
+        builder = AEABuilder.from_aea_project(self._get_cwd())
+        skill_context = SkillContext()
+        behaviour = BuyerBehaviour(name="behaviour", skill_context=skill_context)
+        handler = BuyerHandler(name="handler", skill_context=skill_context)
+        test_skill = Skill(
+            SkillConfig(name="test_skill", author="fetchai"),
+            skill_context=skill_context,
+            handlers={"handler": handler},
+            behaviours={"behaviour": behaviour},
+        )
+        builder.add_component_instance(test_skill)
+        with cd(self._get_cwd()):
+            agent = builder.build()
+            agent.context.shared_state["dialogues"] = BuyerDialogues(
+                agent.context.address
+            )
```

Comment:
Instantiation is responsibility of the skill, not here.

Reply:
skill is creted programmatically and has no access to agent context till agent created by builder

Reply:
updated